core/payment-driver/src/gnt/common.rs

```diff
@@ -0,0 +1,165 @@
+use crate::gnt::config;
+use crate::{utils, PaymentDriverError, PaymentDriverResult};
+use ethereum_types::{Address, H256, U256, U64};
+use futures3::compat::*;
+use web3::contract::{Contract, Options};
+use web3::transports::Http;
+use web3::types::{Bytes, Log, TransactionReceipt};
+use web3::Transport;
+use ya_core_model::driver::{Balance, Currency, PaymentDetails};
+
+pub(crate) fn prepare_gnt_contract(
+    ethereum_client: &EthereumClient,
+    env: &config::EnvConfiguration,
+) -> PaymentDriverResult<Contract<Http>> {
+    prepare_contract(
+        ethereum_client,
+        env.gnt_contract_address,
+        include_bytes!("../contracts/gnt2.json"),
+    )
+}
+
+fn prepare_contract(
+    ethereum_client: &EthereumClient,
+    address: Address,
+    json_abi: &[u8],
+) -> PaymentDriverResult<Contract<Http>> {
+    let contract = ethereum_client.get_contract(address, json_abi)?;
+    Ok(contract)
+}
+
+pub(crate) fn prepare_gnt_faucet_contract(
+    ethereum_client: &EthereumClient,
+    env: &config::EnvConfiguration,
+) -> PaymentDriverResult<Option<Contract<Http>>> {
+    if let Some(gnt_faucet_address) = env.gnt_faucet_address {
+        Ok(Some(prepare_contract(
+            ethereum_client,
+            gnt_faucet_address,
+            include_bytes!("../contracts/faucet.json"),
+        )?))
+    } else {
+        Ok(None)
+    }
+}
+
+pub(crate) async fn get_eth_balance(
+    ethereum_client: &EthereumClient,
+    address: Address,
+) -> PaymentDriverResult<Balance> {
+    let block_number = None;
+    let amount = ethereum_client
+        .get_eth_balance(address, block_number)
+        .await?;
+    Ok(Balance::new(
+        utils::u256_to_big_dec(amount)?,
+        Currency::Eth {},
+    ))
+}
+
+pub(crate) async fn get_gnt_balance(
+    gnt_contract: &Contract<Http>,
+    address: Address,
+) -> PaymentDriverResult<Balance> {
+    gnt_contract
+        .query("balanceOf", (address,), None, Options::default(), None)
+        .compat()
+        .await
+        .map_or_else(
+            |e| Err(PaymentDriverError::LibraryError(format!("{:?}", e))),
+            |amount| {
+                Ok(Balance::new(
+                    utils::u256_to_big_dec(amount)?,
+                    Currency::Gnt {},
+                ))
+            },
+        )
+}
+
+pub(crate) fn verify_gnt_tx<T: Transport>(
+    receipt: &TransactionReceipt,
+    contract: &Contract<T>,
+) -> PaymentDriverResult<()> {
+    verify_gnt_tx_logs(&receipt.logs, contract)?;
+    verify_gnt_tx_status(&receipt.status)?;
+    Ok(())
+}
+
+fn verify_gnt_tx_status(status: &Option<U64>) -> PaymentDriverResult<()> {
+    match status {
+        None => Err(PaymentDriverError::UnknownTransaction),
+        Some(status) => {
+            if *status == U64::from(config::ETH_TX_SUCCESS) {
+                Ok(())
+            } else {
+                Err(PaymentDriverError::FailedTransaction)
+            }
+        }
+    }
+}
+
+fn verify_gnt_tx_logs<T: Transport>(
```

Comment:
why is this its own function? 
- it seems `verify_gnt_tx_logs` is only called once 
- its name is too close for comfort to `verify_gnt_tx_log `

Reply:
I also don't see much benefit in splitting the verification into multiple functions. All of the are called just once and have 2-3 statements.

Reply:
I like to split parsing/verification into smaller functions that can be easily tested.